### DISCUSSION TITLE
How does Micronaut validate native image compatibility of its libraries?

### DISCUSSION THREAD

**Discussion by dzou** [micronaut-projects/micronaut-core#6569] [Q&A] [answered]
Hey, I wanted to ask what kind of tests Micronaut runs to validate its compatibility with native image? I'm interested in finding some examples to learn what the best practices are and how to get more code coverage.

**Comment by ilopmar** [comment, marked as answer]
We run full integration tests every night with different versions of Micronaut and GraalVM. By integration tests I mean:
- Build GraalVM from source code or use latest stable GraalVM version.
- Build native images of applications that we now they work with Micronaut. We have tests for a variety of things: Micronaut Data, RabbitMQ, MQTT, Kafka, Views, Security, OpenAPI,...
- Start those native images, send some curl requests and check the result is the expected.

Our test applications are here https://github.com/orgs/micronaut-graal-tests and this is our CI https://gitlab.com/micronaut-projects/micronaut-graal-tests/-/pipelines.

At this moment we run tests for:
- Micronaut 3.2.x and GraalVM 21.3.0
- Micronaut 3.3.x and GraalVM 21.3.0
- Micronaut 3.3.x and GraalVM 22.0.0-dev (compiled from source code)

**Reply by dzou** [reply]
Thanks for your thorough response.
